File: Leetcode/23. Merge k Sorted Lists/MergeKSortedLists.java
// Leetcode 23. Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/description/
// Runtime: 115ms (very slow...)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
 
 // the solution was not polished and very naive, even used bubble sort
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (null == lists || lists.length == 0) {
            return null;
        }

        // for head
        ListNode head = null;

        // for connecting all chains
        ListNode conn = new ListNode(-1);
        for (int i = 0; i < lists.length; i++) {
            // check the empty chain
            if (lists[i] == null) {
                continue;
            }

            if (head == null) {
                head = lists[i];
            }

            conn.next = lists[i];
            while (null != conn.next) {
                conn = conn.next;
            }
        }

        ListNode rear = head; // i
        ListNode front = null; // i+1

        while (null != rear) {
            front = rear.next;
            while (null != front) {
                if (rear.val > front.val) {
                    int temp = rear.val;
                    rear.val = front.val;
                    front.val = temp;
                }
                front = front.next;
            }
            rear = rear.next;
        }

        return head;
    }
}

File: Leetcode/023. Merge k Sorted Lists/merge_k_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// O(1) space complexity
// main logic is same as merge sort
class Solution
{
public:
    ListNode* mergeKLists(vector<ListNode*>& lists)
    {
        if (lists.empty())
        {
            return nullptr;
        }
        
        ListNode* head = new ListNode(-1);
        head->next = lists[0];
        ListNode* prev = new ListNode(-1);
        ListNode* follow = new ListNode(-1);
        
        for (int i = 1; i < lists.size(); i++)
        {
            prev = head->next;
            follow = lists[i];
            if (follow == nullptr)
            {
                continue;
            }
            
            if (prev == nullptr)
            {
                head->next = follow;
                continue;
            }
            
            // redirect head->next to node with less value
            if (prev->val <= follow->val)
            {
                head->next = prev;
            }
            else
            {
                head->next = follow;
            }
            
            ListNode* p = head;
            
            while (prev && follow)
            {
                if (prev->val <= follow->val)
                {
                    p->next = prev;
                    prev = prev->next;
                }
                else
                {
                    p->next = follow;
                    follow = follow->next;
                }
                p = p->next;
            }
            
            while (prev)
            {
                p->next = prev;
                p = p->next;
                prev = prev->next;
            }
            
            while (follow)
            {
                p->next = follow;
                p = p->next;
                follow = follow->next;
            }
        }
        
        return head->next;
    }
};
